kafka.go

```diff
@@ -194,84 +189,141 @@ func (kafka *kafkaReader) Seek(ctx context.Context, offset int64) (int64, error)
 }
 
-// Internal asynchronous task to fetch messages from Kafka and send to an internal
-// channel.
-func (kafka *kafkaReader) fetchMessagesAsync(ctx context.Context, eventsCh chan<- Message, errorsCh chan<- error) {
-	defer kafka.asyncWait.Done()
+func (kafka *kafkaReader) getMessages(ctx context.Context, offset int64) (*sarama.FetchResponse, error) {
+	errs := make(chan error)
+	val := make(chan *sarama.FetchResponse)
 
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		default:
-		}
-
-		// Find the broker that is the given partition's leader. Failure to fetch the leader is either
-		// the result of an invalid topic/partition OR the broker/leader is unavailable. This can happen
-		// due to a leader election happening (and thus the leader has changed).
-		broker, err := kafka.client.Leader(kafka.topic, kafka.partition)
-		if err != nil {
-			errorsCh <- errors.Wrap(err, "failed to find leader for topic/partition")
-			continue
-		}
+	broker, err := kafka.getLeader(ctx)
+	if err != nil {
+		return nil, err
+	}
 
+	go func() {
 		// The request will wait at most `maxWaitMs` (milliseconds) OR at most `minBytes`,
 		// which ever happens first.
 		request := sarama.FetchRequest{
 			MaxWaitTime: int32(kafka.maxWaitTime.Seconds() / 1000),
 			MinBytes:    int32(kafka.minBytes),
 		}
 
-		offset := atomic.LoadInt64(&kafka.offset)
-
 		request.AddBlock(kafka.topic, kafka.partition, offset, int32(kafka.maxBytes))
+
 		res, err := broker.Fetch(&request)
 		if err != nil {
-			errorsCh <- errors.Wrap(err, "kafka reader failed to fetch a block")
-			continue
+			errs <- errors.Wrap(err, "kafka reader failed to fetch a block")
+		} else {
+			val <- res
 		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errs:
+		return nil, err
+	case res := <-val:
+		return res, nil
+	}
+}
 
-		select {
-		case <-ctx.Done():
+// Find the broker that is the given partition's leader. Failure to fetch the leader is either
+// the result of an invalid topic/partition OR the broker/leader is unavailable. This can happen
+// due to a leader election happening (and thus the leader has changed).
+func (kafka *kafkaReader) getLeader(ctx context.Context) (*sarama.Broker, error) {
+	errs := make(chan error)
+	val := make(chan *sarama.Broker)
+
+	go func() {
+		broker, err := kafka.client.Leader(kafka.topic, kafka.partition)
+		if err != nil {
+			errs <- errors.Wrap(err, "failed to find leader for topic/partition")
+		} else {
+			val <- broker
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errs:
+		return nil, err
+	case broker := <-val:
+		return broker, nil
+	}
+}
+
+// Internal asynchronous task to fetch messages from Kafka and send to an internal
+// channel.
+func (kafka *kafkaReader) fetchMessagesAsync(ctx context.Context, eventsCh chan<- Message, errorsCh chan<- error) {
+	defer kafka.asyncWait.Done()
+
+	for {
+		offset := atomic.LoadInt64(&kafka.offset)
```

Comment:
if we're moving the offset increment to the `Read` method then we'll need to keep this here so that we always load the correct offset to read next.